src/fu-engine-integrity.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuEngine"
+
+#include "config.h"
+
+#include "fu-engine-integrity.h"
+
+/* exported for the self tests */
+void
+fu_engine_integrity_add_checksum(GHashTable *self, const gchar *id, const gchar *csum)
+{
+	g_hash_table_insert(self, g_strdup(id), g_strdup(csum));
+}
+
+static void
+fu_engine_integrity_add_measurement(GHashTable *self, const gchar *id, GBytes *blob)
+{
+	g_autofree gchar *csum = g_compute_checksum_for_bytes(G_CHECKSUM_SHA256, blob);
+	fu_engine_integrity_add_checksum(self, id, csum);
+}
+
+static void
+fu_engine_integrity_measure_acpi(GHashTable *self)
+{
+	g_autofree gchar *path = fu_path_from_kind(FU_PATH_KIND_ACPI_TABLES);
+	const gchar *tables[] = {"SLIC", NULL};
+
+	for (guint i = 0; tables[i] != NULL; i++) {
+		g_autofree gchar *fn = g_build_filename(path, tables[i], NULL);
+		g_autoptr(GBytes) blob = NULL;
+
+		blob = fu_bytes_get_contents(fn, NULL);
+		if (blob != NULL && g_bytes_get_size(blob) > 0) {
+			g_autofree gchar *id = g_strdup_printf("ACPI:%s", tables[i]);
+			fu_engine_integrity_add_measurement(self, id, blob);
+		}
+	}
+}
+
+static void
+fu_engine_integrity_measure_uefi(GHashTable *self)
+{
+	struct {
+		const gchar *guid;
+		const gchar *name;
+	} keys[] = {{FU_EFIVAR_GUID_EFI_GLOBAL, "BootOrder"},
+		    {FU_EFIVAR_GUID_EFI_GLOBAL, "BootCurrent"},
+		    {FU_EFIVAR_GUID_EFI_GLOBAL, "KEK"},
+		    {FU_EFIVAR_GUID_EFI_GLOBAL, "PK"},
+		    {FU_EFIVAR_GUID_SECURITY_DATABASE, "db"},
+		    {FU_EFIVAR_GUID_SECURITY_DATABASE, "dbx"},
+		    {NULL, NULL}};
```

Comment:
How about having the plugin do the measurement?  That way it's only measure when you do uefi-capsule updates.

Reply:
I think we probably ought to do this for other plugins too -- e.g. flashrom.

Reply:
Then maybe should it just be a libfwupdplugin helper that those plugins that change it can call?  I guess I see it either way.

Reply:
My logic was that the daemon should probably define what "system integrity" was -- rather than the plug-ins deciding for themselves. I guess the PK and KEK could be set from the uefi-pk plugin, but I couldn't split the others up into anywhere sane without adding a lot more wiring -- and I couldn't see what else I'd want to add -- e.g. TPM measurements are not relevant here I think.

Reply:
OK, I guess I'm fine with this.  So basically plugins say they might influence integrity data, and daemon measures to see if data changed.

Reply:
Yup; I'm not against polling the plugins for this data, I just don't think we need the extra complexity. If that changes in the future, I'm happy to add it.